Cannot correctly import a json file with an engine path that contains a special character '/'.

'vkv import' does not have an '-e' flag like the 'vkv export' command does to support KV-engines with special characters.

Am I missing something to properly import files with a engine path that contains '/' in the path?

export.json
```json
{
    "secrets/v1": {
        "test01/": {
            "test02/": {
                "test03": {
                    "key": "test-secret"
                }
            }
        }
    }
}
```

> vkv import -d -p secrets/v1 -f test_staging_v1.json
>reading secrets from test_staging_v1.json
>parsing secrets from JSON
>
>preview:
>
>secrets/v1/
>└── v1
>    └── test01
>        └── test02
>            └── test03
>                └── key=***********
>
>apply changes by using the --force flag

The reason for using import for this, is due to having modfied/removed secrets from the exported file, to import into a new vault environment.

Thanks!

ps this tool is amazing, thanks for the hard work.

HI, i've downloaded and tried the latest 0.7.1 release and there does seem to be an issue with the import function.

> vkv import -p secrets/staging--file=staging_v1.json
>reading secrets from staging_v1.json
>parsing secrets from JSON
>[ERROR] error enabling secret engine "secrets": Error making API request.
>
>Namespace: test/migration
>URL: PUT https://vault/v1/sys/mounts/secrets
>Code: 403. Errors:
>
>* 1 error occurred:
>	* permission denied
>
>.

same error using the '-e' flag as well


but if I use the previous verison 0.7.0 it works as expected

> vkv-070 import -p secrets/staging --file=staging_v1.json
>reading secrets from staging_v1.json
>parsing secrets from JSON
>writing secret "secrets/staging/v1/secret01"
>writing secret "secrets/staging/v1/secret02"
>etc....

Thanks!

--- a/cmd/export.go
+++ b/cmd/export.go
@@ -52,7 +52,7 @@ func NewExportCmd() *cobra.Command {
 		SilenceErrors: true,
 		PreRunE:       o.validateFlags,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			enginePath, _ := utils.HandleEnginePath(o.EnginePath, o.Path)
+			enginePath, subPath := utils.HandleEnginePath(o.EnginePath, o.Path)
 
 			printer = prt.NewSecretPrinter(
 				prt.OnlyKeys(o.OnlyKeys),
@@ -66,16 +66,17 @@ func NewExportCmd() *cobra.Command {
 				prt.ShowVersion(o.ShowVersion),
 				prt.ShowMetadata(o.ShowMetadata),
 				prt.WithHyperLinks(o.WithHyperLink),
-				prt.WithEnginePath(enginePath),
+				prt.WithEnginePath(utils.NormalizePath(enginePath)),
 			)
 
-			// prepare map
-			m, err := o.buildMap()
+			secrets, err := vaultClient.ListRecursive(enginePath, subPath, o.SkipErrors)
 			if err != nil {
 				return err
 			}
 
-			if err := printer.Out(m); err != nil {
+			result := utils.UnflattenMap(utils.NormalizePath(path.Join(enginePath, subPath)), utils.ToMapStringInterface(secrets), o.EnginePath)
+
+			if err := printer.Out(result); err != nil {
 				return err
 			}
 
@@ -86,9 +87,9 @@ func NewExportCmd() *cobra.Command {
 	cmd.Flags().SortFlags = false
 
 	// Input
-	cmd.Flags().StringVarP(&o.Path, "path", "p", o.Path, fmt.Sprintf("KV Engine path (env: %s)", envVarExportPrefix+"_PATH"))
-	cmd.Flags().StringVarP(&o.EnginePath, "engine-path", "e", o.EnginePath, "engine path in case your KV-engine contains special characters such as \"/\", the path value will then be appended if specified (\"<engine-path>/<path>\") (env: VKV_EXPORT_ENGINE_PATH)")
-	cmd.Flags().BoolVar(&o.SkipErrors, "skip-errors", o.SkipErrors, "dont exit on errors (permission denied, deleted secrets) (env: VKV_EXPORT_SKIP_ERRORS)")
+	cmd.Flags().StringVarP(&o.Path, "path", "p", o.Path, "KV Engine path (env: VKV_EXPORT_PATH")
+	cmd.Flags().StringVarP(&o.EnginePath, "engine-path", "e", o.EnginePath, "engine path in case your KV-engine contains special characters such as \"/\", the path (-p) flag will then be appended if specified (\"<engine-path>/<path>\") (env: VKV_EXPORT_ENGINE_PATH)")
+	cmd.Flags().BoolVar(&o.SkipErrors, "skip-errors", o.SkipErrors, "don't exit on errors (permission denied, deleted secrets) (env: VKV_EXPORT_SKIP_ERRORS)")
 
 	// Modify
 	cmd.Flags().BoolVar(&o.OnlyKeys, "only-keys", o.OnlyKeys, "show only keys (env: VKV_EXPORT_ONLY_KEYS)")
@@ -119,9 +120,7 @@ func (o *exportOptions) validateFlags(cmd *cobra.Command, args []string) error {
 	case (o.OnlyKeys && o.ShowValues), (o.OnlyPaths && o.ShowValues), (o.OnlyKeys && o.OnlyPaths):
 		return errInvalidFlagCombination
 	case o.EnginePath == "" && o.Path == "":
-		return errors.New("no KV-paths given. Either --engine-path / -e or --path / -p needs to be specified")
-	case o.EnginePath != "" && o.Path != "":
-		return errors.New("cannot specify both engine-path and path")
+		return errors.New("no KV-paths given. Either --engine-path/-e or --path/-p needs to be specified")
 	case true:
 		switch strings.ToLower(o.FormatString) {
 		case "yaml", "yml":
@@ -171,37 +170,4 @@ func (o *exportOptions) validateFlags(cmd *cobra.Command, args []string) error {
 	}
 
 	return nil
-}
-
-func (o *exportOptions) buildMap() (map[string]interface{}, error) {
-	var isSecretPath bool
-
-	rootPath, subPath := utils.HandleEnginePath(o.EnginePath, o.Path)
-
-	// read recursive all secrets
-	s, err := vaultClient.ListRecursive(rootPath, subPath, o.SkipErrors)
-	if err != nil {
-		return nil, err
-	}
-
-	// check if path is a directory or secret path
-	if _, isSecret := vaultClient.ReadSecrets(rootPath, subPath); isSecret == nil {
-		isSecretPath = true
-	}
-
-	path := path.Join(rootPath, subPath)
-	if o.EnginePath != "" {
-		path = subPath
-	}
-
-	// prepare the output map
-	pathMap := utils.PathMap(path, utils.ToMapStringInterface(s), isSecretPath)
-
-	if o.EnginePath != "" {
-		return map[string]interface{}{
-			o.EnginePath: pathMap,
-		}, nil
-	}
-
-	return pathMap, nil
 }

--- a/cmd/export_test.go
+++ b/cmd/export_test.go
@@ -4,7 +4,6 @@ import (
 	"bytes"
 	"io"
 
-	"github.com/FalcoSuessgott/vkv/pkg/fs"
 	prt "github.com/FalcoSuessgott/vkv/pkg/printer/secret"
 )
 
@@ -15,16 +14,6 @@ func (s *VaultSuite) TestValidateExportFlags() {
 		opts *importOptions
 		err  bool
 	}{
-		{
-			name: "path and engine path mutually exclusive",
-			args: []string{"--path=p", "--engine-path=e"},
-			err:  true,
-		},
-		{
-			name: "path or engine path required",
-			args: []string{"--show-values"},
-			err:  true,
-		},
 		{
 			name: "only keys and only paths mutually exclusive",
 			args: []string{"-p=1", "--only-keys", "--only-paths"},
@@ -55,33 +44,72 @@ func (s *VaultSuite) TestValidateExportFlags() {
 func (s *VaultSuite) TestExportImportCommand() {
 	testCases := []struct {
 		name          string
-		path          string
 		expected      string
 		err           bool
 		importCmdArgs []string
 		exportCmdArgs []string
 	}{
 		{
-			name:          "yaml",
-			path:          "yaml",
-			importCmdArgs: []string{"-f=testdata/1.yaml", "-p=yaml"},
+			name:          "import secrets, export from path",
+			importCmdArgs: []string{"-f=testdata/1.yaml"},
 			exportCmdArgs: []string{"-p=yaml", "-f=yaml", "--show-values"},
-			expected:      "testdata/1.yaml",
+			expected: `yaml/:
+  secret:
+    user: password
+`,
 		},
 		{
-			name:          "json",
-			path:          "json",
-			importCmdArgs: []string{"-f=testdata/2.json", "-p=json"},
-			exportCmdArgs: []string{"-p=json", "-f=json", "--show-values"},
-			expected:      "testdata/2.json",
+			name:          "import secrets, overwrite path, read from path",
+			importCmdArgs: []string{"-f=testdata/1.yaml", "-p=yaml2"},
+			exportCmdArgs: []string{"-p=yaml2", "-f=yaml", "--show-values"},
+			expected: `yaml2/:
+  secret:
+    user: password
+`,
 		},
 		{
-			name:          "dryrun",
-			path:          "json",
+			name:          "import secrets, overwrite path and subpath, read from path",
+			importCmdArgs: []string{"-f=testdata/1.yaml", "-p=yaml2/sub"},
+			exportCmdArgs: []string{"-p=yaml2", "-f=yaml", "--show-values"},
+			expected: `yaml2/:
+  sub/:
+    secret:
+      user: password
+`,
+		},
+		{
+			name:          "import secrets, overwrite path with engine path, export from engine path",
+			importCmdArgs: []string{"-f=testdata/2.json", "-e=engine/path"},
+			exportCmdArgs: []string{"-e=engine/path", "-f=json", "--show-values"},
+			expected: `{
+  "engine/path/": {
+    "admin": {
+      "sub": "password"
+    }
+  }
+}
+`,
+		},
+		{
+			name:          "import secrets, overwrite path with engine path and subpath, export from engine path",
+			importCmdArgs: []string{"-f=testdata/2.json", "-e=engine/path", "-p=sub"},
+			exportCmdArgs: []string{"-e=engine/path", "-f=json", "--show-values"},
+			expected: `{
+  "engine/path/": {
+    "sub/": {
+      "admin": {
+        "sub": "password"
+      }
+    }
+  }
+}
+`,
+		},
+		{
+			name:          "no output, error, dryrun",
 			err:           true,
 			importCmdArgs: []string{"-f=testdata/2.json", "-d", "-p=json"},
 			exportCmdArgs: []string{"-p=json", "-f=json", "--show-values"},
-			expected:      "testdata/2.json",
 		},
 	}
 
@@ -94,9 +122,9 @@ func (s *VaultSuite) TestExportImportCommand() {
 			importCmd := NewImportCmd()
 			importCmd.SetArgs(tc.importCmdArgs)
 
-			s.Require().NoError(importCmd.Execute(), ("import " + tc.name))
+			s.Require().NoError(importCmd.Execute(), "import "+tc.name)
 
-			// 2. read secrets and assert
+			// 2. read secrets, capture output and assert
 			b := bytes.NewBufferString("")
 			writer = b
 
@@ -105,14 +133,13 @@ func (s *VaultSuite) TestExportImportCommand() {
 
 			err := exportCmd.Execute()
 
-			s.Require().Equal(tc.err, err != nil, "export "+tc.name)
-
-			// if no error - compare exported secretd with expected value
-			if !tc.err {
+			if tc.err {
+				s.Require().Error(err, "export "+tc.name)
+			} else {
+				// if no error - compare exported secrets with expected value
 				out, _ := io.ReadAll(b)
-				exp, _ := fs.ReadFile(tc.expected)
 
-				s.Require().Equal(string(exp), string(out), "secrets "+tc.name)
+				s.Require().Equal(tc.expected, string(out), "secrets "+tc.name)
 			}
 		})
 	}

--- a/cmd/import.go
+++ b/cmd/import.go
@@ -6,6 +6,7 @@ import (
 	"io"
 	"log"
 	"path"
+	"strings"
 
 	"github.com/FalcoSuessgott/vkv/pkg/fs"
 	prt "github.com/FalcoSuessgott/vkv/pkg/printer/secret"
@@ -14,19 +15,22 @@ import (
 )
 
 type importOptions struct {
-	Force          bool   `env:"FORCE"`
-	DryRun         bool   `env:"DRY_RUN"`
-	Silent         bool   `env:"SILENT"`
-	ShowValues     bool   `env:"SHOW_VALUES"`
-	MaxValueLength int    `env:"MAX_VALUE_LENGTH" envDefault:"12"`
-	Path           string `env:"PATH"`
-	File           string `env:"FILE"`
+	EnginePath string `env:"ENGINE_PATH"`
+	Path       string `env:"PATH"`
+
+	File string `env:"FILE"`
+
+	Force          bool `env:"FORCE"`
+	DryRun         bool `env:"DRY_RUN"`
+	Silent         bool `env:"SILENT"`
+	ShowValues     bool `env:"SHOW_VALUES"`
+	MaxValueLength int  `env:"MAX_VALUE_LENGTH" envDefault:"12"`
 
 	input io.Reader
 }
 
 // NewImportCmd import subcommand.
-// nolint: cyclop, gocognit
+// nolint: cyclop, gocognit, lll
 func NewImportCmd() *cobra.Command {
 	o := &importOptions{}
 
@@ -41,17 +45,6 @@ func NewImportCmd() *cobra.Command {
 		SilenceErrors: true,
 		PreRunE:       o.validateFlags,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			printer = prt.NewSecretPrinter(
-				prt.CustomValueLength(o.MaxValueLength),
-				prt.ShowValues(o.ShowValues),
-				prt.ToFormat(prt.Base),
-				prt.WithVaultClient(vaultClient),
-				prt.WithWriter(writer),
-				prt.ShowVersion(true),
-				prt.ShowMetadata(true),
-				prt.WithEnginePath(o.Path),
-			)
-
 			// get user input via -f or STDIN
 			input, err := o.getInput()
 			if err != nil {
@@ -64,24 +57,71 @@ func NewImportCmd() *cobra.Command {
 				return err
 			}
 
+			// if no path specified, use the path from the secrets to be imported
+			if o.EnginePath == "" && o.Path == "" {
+				fmt.Println("no path specified, trying to determine root path from the provided input")
+
+				rootPath, err := utils.GetRootElement(secrets)
+				if err != nil {
+					return fmt.Errorf("try specifying a destination path using -p/-e. %w", err)
+				}
+
+				fmt.Printf("using \"%s\" as KV engine path\n", rootPath)
+
+				// detect whether its an engine path or a normal root path
+				if len(strings.Split(rootPath, utils.Delimiter)) > 1 {
+					o.EnginePath = rootPath
+				} else {
+					o.Path = rootPath
+				}
+			}
+
+			// read existing secrets from the rootPath
+			rootPath, subPath := utils.HandleEnginePath(o.EnginePath, o.Path)
+
+			printer = prt.NewSecretPrinter(
+				prt.CustomValueLength(o.MaxValueLength),
+				prt.ShowValues(o.ShowValues),
+				prt.ToFormat(prt.Base),
+				prt.WithVaultClient(vaultClient),
+				prt.WithWriter(writer),
+				prt.ShowVersion(true),
+				prt.ShowMetadata(true),
+				prt.ShowVersion(false),
+				prt.WithEnginePath(utils.NormalizePath(rootPath)),
+			)
+
 			// print preview during dryrun and exit
 			if o.DryRun {
-				return o.dryRun(secrets)
+				// replace the root path in the secrets with the specified path
+				secretsWithNewPath := make(map[string]interface{})
+				for _, v := range secrets {
+					secretsWithNewPath = utils.UnflattenMap(utils.NormalizePath(path.Join(rootPath, subPath)), utils.ToMapStringInterface(v), o.EnginePath)
+				}
+
+				return o.dryRun(rootPath, subPath, secretsWithNewPath)
 			}
 
 			// enable kv engine, error if already enabled, unless force is used
-			if err := vaultClient.EnableKV2EngineErrorIfNotForced(o.Force, o.Path); err != nil {
+			if err := vaultClient.EnableKV2EngineErrorIfNotForced(o.Force, rootPath); err != nil {
 				return err
 			}
 
 			// write secrets
-			if err := o.writeSecrets(secrets); err != nil {
+			if err := o.writeSecrets(rootPath, subPath, secrets); err != nil {
 				return err
 			}
 
 			// show result if not silence mode
 			if !o.Silent {
-				return o.printResult()
+				result, err := o.printResult(rootPath, subPath)
+				if err != nil {
+					return err
+				}
+
+				if err := printer.Out(result); err != nil {
+					return err
+				}
 			}
 
 			return nil
@@ -90,6 +130,7 @@ func NewImportCmd() *cobra.Command {
 
 	// Input
 	cmd.Flags().StringVarP(&o.Path, "path", "p", o.Path, "KVv2 Engine path (env: VKV_IMPORT_PATH)")
+	cmd.Flags().StringVarP(&o.EnginePath, "engine-path", "e", o.EnginePath, "engine path in case your KV-engine contains special characters such as \"/\", the path (-p) flag will then be appended if specified (\"<engine-path>/<path>\") (env: VKV_IMPORT_PATH)")
 	cmd.Flags().StringVarP(&o.File, "file", "f", o.File, "path to a file containing vkv yaml or json output (env: VKV_IMPORT_FILE)")
 
 	// Options
@@ -100,11 +141,6 @@ func NewImportCmd() *cobra.Command {
 	cmd.Flags().IntVar(&o.MaxValueLength, "max-value-length", o.MaxValueLength, "maximum char length of values. Set to \"-1\" for disabling "+
 		"(env: VKV_IMPORT_MAX_VALUE_LENGTH)")
 
-	// flag constraints
-	if err := cmd.MarkFlagRequired("path"); err != nil {
-		log.Fatal(err)
-	}
-
 	o.input = cmd.InOrStdin()
 
 	return cmd
@@ -112,8 +148,6 @@ func NewImportCmd() *cobra.Command {
 
 func (o *importOptions) validateFlags(cmd *cobra.Command, args []string) error {
 	switch {
-	case len(args) == 0 && o.Path == "":
-		return errors.New("no KV-path given, -path / -p needs to be specified")
 	case o.Force && o.DryRun:
 		return fmt.Errorf("%w: %s", errInvalidFlagCombination, "cannot specify both --force and --dry-run")
 	case o.Silent && o.DryRun:
@@ -171,25 +205,30 @@ func (o *importOptions) parseInput(input []byte) (map[string]interface{}, error)
 	return json, nil
 }
 
-func (o *importOptions) writeSecrets(secrets map[string]interface{}) error {
+func (o *importOptions) writeSecrets(rootPath, subPath string, secrets map[string]interface{}) error {
 	transformedMap := make(map[string]interface{})
 	utils.FlattenMap(secrets, transformedMap, "")
 
 	for p, m := range transformedMap {
-		secrets, ok := m.(map[string]interface{})
+		secret, ok := m.(map[string]interface{})
 		if !ok {
-			log.Fatalf("cannot convert %T to map[string]interface", secrets)
+			log.Fatalf("cannot convert %T to map[string]interface", secret)
 		}
 
-		rootPath, subPath := utils.SplitPath(o.Path)
-		_, subPath2 := utils.SplitPath(p)
+		// replace original path with the new engine path
+		t, _ := utils.GetRootElement(secrets)
+		newSubPath := strings.TrimPrefix(p, t)
 
-		nSubPath := path.Join(subPath, subPath2)
-		if err := vaultClient.WriteSecrets(rootPath, nSubPath, secrets); err != nil {
+		// unless a subpath has been specified by the user
+		if subPath != "" {
+			newSubPath = path.Join(subPath, newSubPath)
+		}
+
+		if err := vaultClient.WriteSecrets(rootPath, newSubPath, secret); err != nil {
 			return fmt.Errorf("error writing secret \"%s\": %w", p, err)
 		}
 
-		fmt.Fprintf(writer, "writing secret \"%s\" \n", path.Join(rootPath, nSubPath))
+		fmt.Fprintf(writer, "writing secret \"%s\" \n", path.Join(rootPath, newSubPath))
 	}
 
 	fmt.Fprintln(writer, "successfully imported all secrets")
@@ -197,29 +236,39 @@ func (o *importOptions) writeSecrets(secrets map[string]interface{}) error {
 	return nil
 }
 
-func (o *importOptions) dryRun(secrets map[string]interface{}) error {
+func (o *importOptions) dryRun(rootPath, subPath string, secrets map[string]interface{}) error {
+	fmt.Printf("fetching KV secrets from \"%s\" (if any)\n", utils.NormalizePath(path.Join(rootPath, subPath)))
+
+	tmp, err := vaultClient.ListRecursive(rootPath, subPath, true)
+	if err != nil {
+		return fmt.Errorf("error listing secrets from \"%s/%s\": %w", rootPath, subPath, err)
+	}
+
+	if len(utils.ToMapStringInterface(tmp)) == 0 {
+		fmt.Println("no secrets found - nothing to compare with")
+	}
+
+	existingSecrets := utils.UnflattenMap(utils.NormalizePath(path.Join(rootPath, subPath)), utils.ToMapStringInterface(tmp), o.EnginePath)
+
+	// check whether new and existing secrets are equal
+	if fmt.Sprint(secrets) == fmt.Sprint(existingSecrets) {
+		fmt.Fprintln(writer, "")
+		fmt.Fprintln(writer, "input matches secrets - no changes needed:")
+		fmt.Fprintln(writer, "")
+
+		if err := printer.Out(existingSecrets); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	fmt.Fprintf(writer, "deep merging provided secrets with existing secrets read from \"%s\"\n", utils.NormalizePath(path.Join(rootPath, subPath)))
 	fmt.Fprintln(writer, "")
 	fmt.Fprintln(writer, "preview:")
 	fmt.Fprintln(writer, "")
 
-	// read existing ecrets from the rootPath
-	rootPath, _ := utils.SplitPath(o.Path)
-	existingSecrets := make(map[string]interface{})
-
-	tmp, err := vaultClient.ListRecursive(rootPath, "", false)
-	if err == nil {
-		existingSecrets = utils.PathMap(rootPath, utils.ToMapStringInterface(tmp), false)
-	}
-
-	// add new secrets to it
-	newSecrets := make(map[string]interface{})
-	for _, v := range secrets {
-		newSecrets = utils.PathMap(o.Path, utils.ToMapStringInterface(v), false)
-	}
-
-	// deep merge both secrets
-	mergedSecrets := utils.DeepMergeMaps(newSecrets, existingSecrets)
-	if err := printer.Out(mergedSecrets); err != nil {
+	if err := printer.Out(utils.DeepMergeMaps(secrets, existingSecrets)); err != nil {
 		return err
 	}
 
@@ -229,23 +278,27 @@ func (o *importOptions) dryRun(secrets map[string]interface{}) error {
 	return nil
 }
 
-func (o *importOptions) printResult() error {
+func (o *importOptions) printResult(rootPath, subPath string) (map[string]interface{}, error) {
 	fmt.Fprintln(writer, "")
 	fmt.Fprintln(writer, "result:")
 	fmt.Fprintln(writer, "")
 
-	rootPath, _ := utils.SplitPath(o.Path)
+	printer = prt.NewSecretPrinter(
+		prt.CustomValueLength(o.MaxValueLength),
+		prt.ShowValues(o.ShowValues),
+		prt.ToFormat(prt.Base),
+		prt.WithVaultClient(vaultClient),
+		prt.WithWriter(writer),
+		prt.ShowVersion(true),
+		prt.ShowMetadata(true),
+		prt.ShowVersion(true),
+		prt.WithEnginePath(utils.NormalizePath(rootPath)),
+	)
 
-	s, err := vaultClient.ListRecursive(rootPath, "", false)
+	secrets, err := vaultClient.ListRecursive(rootPath, subPath, false)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	secrets := utils.PathMap(rootPath, utils.ToMapStringInterface(s), false)
-
-	if err := printer.Out(secrets); err != nil {
-		return err
-	}
-
-	return nil
+	return utils.UnflattenMap(utils.NormalizePath(path.Join(rootPath, subPath)), utils.ToMapStringInterface(secrets), o.EnginePath), nil
 }

--- a/cmd/import_test.go
+++ b/cmd/import_test.go
@@ -20,19 +20,22 @@ func (s *VaultSuite) TestValidateImportFlags() {
 			err:  true,
 		},
 		{
-			name: "no path should fail",
-			err:  true,
-			args: []string{},
-		},
-		{
 			name: "silent and dry-run should fail",
 			err:  true,
 			args: []string{"-p=o", "-s", "-d"},
+		},
+		{
+			name: "file and STDIN should fail",
+			err:  true,
+			args: []string{"-f=sss", "-"},
 		},
 	}
 
 	for _, tc := range testCases {
 		cmd := NewImportCmd()
+
+		writer = io.Discard
+
 		cmd.SetArgs(tc.args)
 
 		err := cmd.Execute()
@@ -62,7 +65,6 @@ func (s *VaultSuite) TestGetInput() {
 			expected: `yaml/:
   secret:
     user: password
-
 `,
 		},
 		{

--- a/cmd/server.go
+++ b/cmd/server.go
@@ -75,8 +75,6 @@ func (o *serverOptions) validateFlags(cmd *cobra.Command, args []string) error {
 }
 
 func (o *serverOptions) buildMap() (map[string]interface{}, error) {
-	var isSecretPath bool
-
 	rootPath, subPath := utils.HandleEnginePath(o.EnginePath, o.Path)
 
 	// read recursive all secrets
@@ -85,18 +83,13 @@ func (o *serverOptions) buildMap() (map[string]interface{}, error) {
 		return nil, err
 	}
 
-	// check if path is a directory or secret path
-	if _, isSecret := vaultClient.ReadSecrets(rootPath, subPath); isSecret == nil {
-		isSecretPath = true
-	}
-
 	path := path.Join(rootPath, subPath)
 	if o.EnginePath != "" {
 		path = subPath
 	}
 
 	// prepare the output map
-	pathMap := utils.PathMap(path, utils.ToMapStringInterface(s), isSecretPath)
+	pathMap := utils.UnflattenMap(path, utils.ToMapStringInterface(s))
 
 	if o.EnginePath != "" {
 		return map[string]interface{}{

--- a/cmd/testdata/1.yaml
+++ b/cmd/testdata/1.yaml
@@ -1,4 +1,3 @@
 yaml/:
   secret:
     user: password
-

--- a/cmd/testdata/3.json
+++ b/cmd/testdata/3.json
@@ -1,0 +1,7 @@
+{
+  "a/b/c/": {
+    "admin": {
+      "sub": "password"
+    }
+  }
+}

--- a/cmd/testdata/secret.yaml
+++ b/cmd/testdata/secret.yaml
@@ -1,22 +1,23 @@
 {
-  "admin": {
-    "sub": "password"
-  },
-  "demo": {
-    "foo": "bar"
-  },
-  "sub/": {
-    "demo": {
-      "demo": "hello world",
-      "password": "s3cre5",
-      "user": "admin"
+  "secret/": {
+    "admin": {
+      "sub": "password"
     },
-    "sub2/": {
+    "demo": {
+      "foo": "bar"
+    },
+    "sub/": {
       "demo": {
-        "admin": "key",
-        "foo": "bar",
-        "password": "password",
-        "user": "user"
+        "demo": "hello world",
+        "password": "s3cre5<",
+        "user": "admin"
+      },
+      "sub2/": {
+        "demo": {
+          "foo": "bar",
+          "password": "password",
+          "user": "user"
+        }
       }
     }
   }

--- a/docs/export.md
+++ b/docs/export.md
@@ -3,6 +3,22 @@
 
 See the [CLI Reference](https://falcosuessgott.github.io/vkv/cmd/vkv_export/) for more details on the supported flags and env vars.
 
+!!! warning
+    Vault allows `/` in the name of a KV engine. This makes it difficult for `vkv` to distinguish between directories and the KV engine name..
+    
+    If your KV engine name/mount contains a `/` you have to specify it using `--engine-path|-e`, otherwise `vkv` will output the secrets wrong.
+
+    This also applies for any `vkv import ...` operations.
+
+!!! info
+    `vkv` handles 3 different path arguments, specified using `-e|-p`
+
+    1. `root path`: any normal KV mount. Use `-p`.
+    2. `engine-path`: in case your KV mount contains a `/`. Use `-e`.
+    3. `sub path`: the path to the corresponding directory within a KV mount. 
+    When using `-p` this is everything after the first `/`: e.g: `kv/prod/db/`; root path=`kv`, subpath=`prod/db`. 
+    In conjunction with a `-e` you can specify a sub-path by using -p: `-e=kv/prod -p=db`.
+ 
 ## base
 ```bash
 > vkv export -p secret -f=base               

--- a/docs/import.md
+++ b/docs/import.md
@@ -1,8 +1,28 @@
 # Import
 
-`vkv import` requires an engine path (`--path`) and will accepts `vkv`s YAML or JSON output (`vkv export -f=yaml|json`) either by invoking `vkv import -` (for STDIN) or by specifying a file (`--file`). `vkv` will create the specified path if the engine does not exist yet and will error if it does, unless `--force` is specified. 
+`vkv import` tries to determine the path from the input, if that is not possible you can specify on using `--path|--engine-path`. 
+
+`vkv import` accepts `vkv`s YAML or JSON output (`vkv export -f=yaml|json`) either by invoking `vkv import -` (for STDIN) or by specifying a file (`--file`). 
+
+`vkv` will create the specified path if the engine does not exist yet and will error if it does, unless `--force` is specified. 
 
 See the [CLI Reference](https://github.com/FalcoSuessgott/vkv/cmd/vkv_import/) for more details on the supported flags and env vars.
+
+!!! warning
+    Vault allows `/` in the name of KV engine. This makes it difficult for `vkv` to distinguish between directories and the KV engine name..
+    
+    If your KV engine name/mount contains a `/` you have to specify it using `--engine-path|-e`, otherwise `vkv` will output the secrets wrong.
+
+    This also applies for any `vkv import ...` operations.
+
+!!! info
+    `vkv` handles 3 different path arguments, specified using `-e|-p`
+
+    1. `root path`: any normal KV mount. Use `-p`.
+    2. `engine-path`: in case your KV mount contains a `/`. Use `-e`.
+    3. `sub path`: the path to the corresponding directory within a KV mount. 
+    When using `-p` this is everything after the first `/`: e.g: `kv/prod/db/`; root path=`kv`, subpath=`prod/db`. 
+    In conjunction with a `-e` you can specify a sub-path by using -p: `-e=kv/prod -p=db`.
 
 ## Example Usage
 ```bash

--- a/go.mod
+++ b/go.mod
@@ -10,6 +10,7 @@ require (
 	github.com/gin-gonic/gin v1.10.0
 	github.com/hashicorp/vault/api v1.14.0
 	github.com/juju/ansiterm v1.0.0
+	github.com/k0kubun/pp/v3 v3.2.0
 	github.com/muesli/mango-cobra v1.2.0
 	github.com/muesli/roff v0.1.0
 	github.com/savioxavier/termlink v1.3.0

--- a/pkg/printer/secret/base.go
+++ b/pkg/printer/secret/base.go
@@ -17,12 +17,12 @@ func (p *Printer) printBase(secrets map[string]interface{}) error {
 	m := make(map[string]interface{})
 
 	for _, k := range utils.SortMapKeys(secrets) {
-		baseName := p.enginePath + utils.Delimiter
+		baseName := p.enginePath
 
 		if p.withHyperLinks {
 			addr := fmt.Sprintf("%s/ui/vault/secrets/%s/kv", p.vaultClient.Client.Address(), p.enginePath)
 
-			baseName = termlink.Link(p.enginePath+utils.Delimiter, addr, false)
+			baseName = termlink.Link(p.enginePath, addr, false)
 		}
 
 		if p.vaultClient != nil {
@@ -60,8 +60,9 @@ func (p *Printer) printTree(rootPath, subPath string, m map[string]interface{}) 
 	//nolint: nestif
 	if strings.HasSuffix(subPath, utils.Delimiter) {
 		for _, i := range utils.SortMapKeys(m) {
-			//nolint: forcetypeassert
-			tree.AddBranch(p.printTree(rootPath, subPath+i, m[i].(map[string]interface{})))
+			if data, ok := m[i].(map[string]interface{}); ok {
+				tree.AddBranch(p.printTree(rootPath, subPath+i, data))
+			}
 		}
 	} else {
 		for _, k := range utils.SortMapKeys(m) {

--- a/pkg/printer/secret/base_test.go
+++ b/pkg/printer/secret/base_test.go
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"testing"
 
+	"github.com/FalcoSuessgott/vkv/pkg/utils"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
@@ -116,14 +117,15 @@ func TestPrintBase(t *testing.T) {
 		var b bytes.Buffer
 		tc.opts = append(tc.opts,
 			WithWriter(&b),
-			WithEnginePath(tc.rootPath),
+			WithEnginePath(utils.NormalizePath(tc.rootPath)),
 		)
 
 		p := NewSecretPrinter(tc.opts...)
 
-		m := map[string]interface{}{}
+		m := map[string]interface{}{
+			utils.NormalizePath(tc.rootPath): tc.s,
+		}
 
-		m[tc.rootPath+"/"] = tc.s
 		require.NoError(t, p.Out(m))
 		assert.Equal(t, tc.output, b.String(), tc.name)
 	}

--- a/pkg/printer/secret/secret_printer.go
+++ b/pkg/printer/secret/secret_printer.go
@@ -79,9 +79,7 @@ func CustomValueLength(length int) Option {
 // OnlyKeys flag for only showing secrets keys.
 func OnlyKeys(b bool) Option {
 	return func(p *Printer) {
-		if b {
-			p.onlyKeys = true
-		}
+		p.onlyKeys = b
 	}
 }
 
@@ -97,9 +95,7 @@ func WithHyperLinks(b bool) Option {
 // OnlyPaths flag for only printing kv paths.
 func OnlyPaths(b bool) Option {
 	return func(p *Printer) {
-		if b {
-			p.onlyPaths = true
-		}
+		p.onlyPaths = b
 	}
 }
 
@@ -117,30 +113,24 @@ func WithWriter(w io.Writer) Option {
 	}
 }
 
-// ShowValues flag for unmasking secrets in output.
+// ShowValues flag for displaying the secrets version.
 func ShowValues(b bool) Option {
 	return func(p *Printer) {
-		if b {
-			p.showValues = true
-		}
+		p.showValues = b
 	}
 }
 
 // ShowVersion flag for unmasking secrets in output.
 func ShowVersion(b bool) Option {
 	return func(p *Printer) {
-		if b {
-			p.showVersion = true
-		}
+		p.showVersion = b
 	}
 }
 
 // ShowMetadata flag for unmasking secrets in output.
 func ShowMetadata(b bool) Option {
 	return func(p *Printer) {
-		if b {
-			p.showMetadata = true
-		}
+		p.showMetadata = b
 	}
 }
 
@@ -191,6 +181,13 @@ func NewSecretPrinter(opts ...Option) *Printer {
 	}
 
 	return p
+}
+
+// Update update printer applies the given options.
+func Update(p *Printer, opts ...Option) {
+	for _, opt := range opts {
+		opt(p)
+	}
 }
 
 // Out prints out the secrets according all configured options.

--- a/pkg/printer/secret/yaml.go
+++ b/pkg/printer/secret/yaml.go
@@ -12,7 +12,7 @@ func (p *Printer) printYAML(secrets map[string]interface{}) error {
 		return err
 	}
 
-	fmt.Fprintln(p.writer, string(out))
+	fmt.Fprint(p.writer, string(out))
 
 	return nil
 }

--- a/pkg/printer/secret/yaml_test.go
+++ b/pkg/printer/secret/yaml_test.go
@@ -35,7 +35,6 @@ func TestPrintYAML(t *testing.T) {
   secret:
     key: value
     user: password
-
 `,
 		},
 		{
@@ -55,7 +54,6 @@ func TestPrintYAML(t *testing.T) {
   secret:
     key: ""
     user: ""
-
 `,
 		},
 	}

--- a/pkg/utils/utils.go
+++ b/pkg/utils/utils.go
@@ -4,6 +4,7 @@ package utils
 import (
 	"bytes"
 	"encoding/json"
+	"errors"
 	"log"
 	"path"
 	"path/filepath"
@@ -22,6 +23,10 @@ const (
 // Keys type for receiving all keys of a map.
 type Keys []string
 
+func NormalizePath(path string) string {
+	return filepath.Clean(path) + Delimiter
+}
+
 // FlattenMap flattens a nested map into a single map with its.
 func FlattenMap(a, b map[string]interface{}, key string) {
 	for k, v := range a {
@@ -35,26 +40,33 @@ func FlattenMap(a, b map[string]interface{}, key string) {
 	}
 }
 
-// PathMap takes a path like "a/b/c" and returns a map like map[a] -> map[b] -> map[c].
-// if isSecretPath is true, then c does not have a / as suffix.
-func PathMap(path string, s map[string]interface{}, isSecretPath bool) map[string]interface{} {
+// UnflattenMap takes a path like "a/b/c" and returns a map like map[a] -> map[b] -> map[c].
+// elements in ignoreElements are not splitted.
+func UnflattenMap(path string, data map[string]interface{}, ignoreElements ...string) map[string]interface{} {
 	m := map[string]interface{}{}
 
 	parts := strings.Split(path, Delimiter)
 
 	if path == "" {
-		return s
+		return data
+	}
+
+	path = NormalizePath(path)
+
+	// if element is to be ignored
+	for _, ignore := range ignoreElements {
+		if strings.HasPrefix(path, NormalizePath(ignore)) {
+			m[NormalizePath(ignore)] = UnflattenMap(strings.TrimPrefix(path, NormalizePath(ignore)), data, ignoreElements...)
+
+			return m
+		}
 	}
 
 	if len(parts) > 1 {
-		m[parts[0]+Delimiter] = PathMap(strings.Join(parts[1:], Delimiter), s, isSecretPath)
+		m[NormalizePath(parts[0])] = UnflattenMap(strings.Join(parts[1:], Delimiter), data, ignoreElements...)
 	} else {
-		// if path leads to a vault kv directory, append a "/"
-		if !isSecretPath {
-			path += Delimiter
-		}
-
-		m[path] = s
+		// or dont if there is no parts left
+		m[strings.TrimSuffix(path, Delimiter)] = data
 	}
 
 	return m
@@ -69,6 +81,18 @@ func SplitPath(path string) (string, string) {
 	}
 
 	return strings.Join(parts, Delimiter), ""
+}
+
+func GetRootElement(m map[string]interface{}) (string, error) {
+	if len(m) > 1 {
+		return "", errors.New("map contains multiple keys, cannot detect root element")
+	}
+
+	for k := range m {
+		return k, nil
+	}
+
+	return "", errors.New("this shouldn't happen")
 }
 
 func removeEmptyElements(s []string) []string {

--- a/pkg/utils/utils_test.go
+++ b/pkg/utils/utils_test.go
@@ -7,6 +7,12 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestNormalizePath(t *testing.T) {
+	assert.Equal(t, "a/b/c/", NormalizePath("a/b/c"))
+	assert.Equal(t, "a/b/c/", NormalizePath("a/b/c/"))
+	assert.Equal(t, "a/b/c/", NormalizePath("a/b/c/////"))
+}
+
 func TestRemoveExtension(t *testing.T) {
 	s := "path/to/file.txt"
 	assert.Equal(t, "path/to/file", RemoveExtension(s))
@@ -139,22 +145,39 @@ func TestFlattenMap(t *testing.T) {
 	}
 }
 
-func TestPathMap(t *testing.T) {
+func TestGetRootElement(t *testing.T) {
+	m := map[string]interface{}{
+		"k": false,
+	}
+
+	res, err := GetRootElement(m)
+	require.NoError(t, err)
+	assert.Equal(t, "k", res)
+
+	n := map[string]interface{}{
+		"k": false,
+		"b": true,
+	}
+
+	_, err = GetRootElement(n)
+	require.Error(t, err)
+}
+
+func TestUnflattenMap(t *testing.T) {
 	testCases := []struct {
-		name         string
-		path         string
-		m            map[string]interface{}
-		isSecretPath bool
-		expected     map[string]interface{}
+		name           string
+		path           string
+		m              map[string]interface{}
+		ignoreElements []string
+		expected       map[string]interface{}
 	}{
 		{
-			name: "secret path",
+			name: "simple",
 			path: "root/sub",
 			m: map[string]interface{}{
 				"k":  "v",
 				"k2": 12,
 			},
-			isSecretPath: true,
 			expected: map[string]interface{}{
 				"root/": map[string]interface{}{
 					"sub": map[string]interface{}{
@@ -165,36 +188,43 @@ func TestPathMap(t *testing.T) {
 			},
 		},
 		{
-			name:         "directory path",
-			path:         "root/sub",
-			m:            map[string]interface{}{},
-			isSecretPath: false,
+			name: "simple with ignored fields",
+			path: "root/sub",
+			m: map[string]interface{}{
+				"k":  "v",
+				"k2": 12,
+			},
+			ignoreElements: []string{"root/sub"},
 			expected: map[string]interface{}{
-				"root/": map[string]interface{}{
-					"sub/": map[string]interface{}{},
+				"root/sub/": map[string]interface{}{
+					"k":  "v",
+					"k2": 12,
 				},
 			},
 		},
 		{
-			name:         "only root",
-			path:         "root",
-			m:            map[string]interface{}{},
-			isSecretPath: false,
-			expected: map[string]interface{}{
-				"root/": map[string]interface{}{},
+			name: "complex with ignored fields",
+			path: "root/sub/a/b/c",
+			m: map[string]interface{}{
+				"k":  "v",
+				"k2": 12,
 			},
-		},
-		{
-			name:         "only root",
-			path:         "",
-			m:            map[string]interface{}{},
-			isSecretPath: false,
-			expected:     map[string]interface{}{},
+			ignoreElements: []string{"root/sub", "b/c"},
+			expected: map[string]interface{}{
+				"root/sub/": map[string]interface{}{
+					"a/": map[string]interface{}{
+						"b/c/": map[string]interface{}{
+							"k":  "v",
+							"k2": 12,
+						},
+					},
+				},
+			},
 		},
 	}
 
 	for _, tc := range testCases {
-		assert.Equal(t, tc.expected, PathMap(tc.path, tc.m, tc.isSecretPath), tc.name)
+		assert.Equal(t, tc.expected, UnflattenMap(tc.path, tc.m, tc.ignoreElements...), tc.name)
 	}
 }
 
@@ -569,7 +599,7 @@ func TestHandleEnginePath(t *testing.T) {
 			expectedSubPath:  "5/6",
 		},
 		{
-			name:             " only engine path",
+			name:             "only engine path",
 			enginePath:       "1/2/3/4",
 			expectedRootPath: "1/2/3/4",
 		},

--- a/pkg/vault/engine.go
+++ b/pkg/vault/engine.go
@@ -3,9 +3,6 @@ package vault
 import (
 	"fmt"
 	"path"
-	"strings"
-
-	"github.com/FalcoSuessgott/vkv/pkg/utils"
 )
 
 const (
@@ -24,8 +21,13 @@ func (v *Vault) GetEngineDescription(rootPath string) (string, error) {
 	}
 
 	if data != nil {
+		desc, ok := data.Data["description"]
+		if !ok {
+			return "", nil
+		}
+
 		//nolint: forcetypeassert
-		return data.Data["description"].(string), nil
+		return desc.(string), nil
 	}
 
 	return "", fmt.Errorf("could not get engine description for path: \"%s\"", rootPath)
@@ -38,8 +40,22 @@ func (v *Vault) GetEngineTypeVersion(rootPath string) (string, string, error) {
 	}
 
 	if data != nil {
-		//nolint: forcetypeassert
-		return data.Data["type"].(string), data.Data["options"].(map[string]interface{})["version"].(string), nil
+		var eType, eVersion string
+
+		t, ok := data.Data["type"]
+
+		if ok {
+			//nolint: forcetypeassert
+			eType = t.(string)
+		}
+
+		v, ok := data.Data["options"]
+		if ok {
+			//nolint: forcetypeassert
+			eVersion = v.(map[string]interface{})["version"].(string)
+		}
+
+		return eType, eVersion, nil
 	}
 
 	return "", "", fmt.Errorf("could not get engine type for path: \"%s\"", rootPath)
@@ -84,25 +100,35 @@ func (v *Vault) EnableKV1Engine(rootPath string) error {
 // EnableKV2EngineErrorIfNotForced enables a KVv2 Engine and errors if
 // already enabled, unless force is set to true.
 func (v *Vault) EnableKV2EngineErrorIfNotForced(force bool, path string) error {
-	rootPath, _ := utils.SplitPath(path)
-
-	if len(strings.Split(path, utils.Delimiter)) > 1 {
-		//nolint: nilerr
-		if err := v.EnableKV2Engine(rootPath); err != nil {
-			return nil
+	// check if engine exists
+	engineType, kvVersion, err := v.GetEngineTypeVersion(path)
+	// engine does not exists, so we enable it and exit
+	if err != nil {
+		if err := v.EnableKV2Engine(path); err != nil {
+			return fmt.Errorf("error enabling secret engine \"%s\": %w", path, err)
 		}
+
+		return nil
 	}
 
-	if v.EnableKV2Engine(rootPath) != nil && !force {
-		return fmt.Errorf("a secret engine under \"%s\" is already enabled. Use --force for overwriting", rootPath)
+	// engine exists, but is not of type kvv2
+	if err == nil && (engineType != "kv" || kvVersion != "2") {
+		return fmt.Errorf("engine \"%s\" is not of type kv2", path)
 	}
 
-	if err := v.DisableKV2Engine(rootPath); err != nil {
-		return fmt.Errorf("error disabling secret engine \"%s\": %w", rootPath, err)
+	// engine exists but no force flag used for using that engine
+	if err == nil && !force {
+		return fmt.Errorf("a secret engine under \"%s\" is already enabled. Use --force for overwriting", path)
 	}
 
-	if err := v.EnableKV2Engine(rootPath); err != nil {
-		return fmt.Errorf("error enabling secret engine \"%s\": %w", rootPath, err)
+	// force flag is used, so we disable the engine
+	if err := v.DisableKV2Engine(path); err != nil {
+		return fmt.Errorf("error disabling secret engine \"%s\": %w", path, err)
+	}
+
+	// enable the engine
+	if err := v.EnableKV2Engine(path); err != nil {
+		return fmt.Errorf("error enabling secret engine \"%s\": %w", path, err)
 	}
 
 	return nil

--- a/scripts/prepare-vault.sh
+++ b/scripts/prepare-vault.sh
@@ -19,3 +19,14 @@ vault kv put secret_2/demo foo=bar
 vault kv put secret_2/admin sub=password
 vault kv put secret_2/sub/demo demo="hello world" user=admin password='s3cre5<'
 vault kv put secret_2/sub/sub2/demo foo=bar-updated user=user password=password
+
+
+# # test cases
+# 1. rootpath -> rootpath
+# 2. subpath -> rootpath
+# 3. subpath -> subpath
+# 4. root path -> enginepath
+# 5. enginepath -> enginepath
+# 6. engie path + sub > engine path + sub
+# 7. engine path -> root path
+# 8. engine path -> subpath

